### PR TITLE
feat: add private meeting room booking skill

### PR DESCRIPTION
**Pull request by Light-code11**
## Summary

- add a dedicated meeting-room skill for availability, booking, listing, and cancellation
- distinguish hourly room requests from full-day coworking requests
- resolve dates and times in Melbourne time and require a private confirmation before charging
- show unavailable intervals for a requested date and explicitly confirm that every unlisted time is currently available
- bind signed, ten-minute confirmation and cancellation actions to the verified Slack actor
- deliver room, booking, balance, and cancellation details by DM, with public-channel privacy fallbacks
- handle duplicate events/clicks safely through stable request IDs and backend idempotency
- run the meeting-room tests in the pull-request security gate
- keep the router catalog below its enforced token budget after integrating current `main`

## Scope

V1 supports the single MLAI Meeting Room and self-booking only. It does not add attendees, meeting titles, recurrence, calendar invitations, reminders, admin booking for others, or an external calendar. Members can ask Roo about a specific interval or a whole date; date-wide results list busy times and state that all other times are available.

The feature is disabled by default with `MEETING_ROOM_BOOKING_ENABLED`.

## Validation

- `pytest roo/tests/test_meeting_room_booking.py roo/tests/test_router_catalog.py roo/tests/test_routing_eval_gate.py roo/tests/test_router_v2.py` (38 passed)
- PR security test selection, including meeting-room coverage (511 passed)
- deterministic routing evaluation (249 cases, 0 misroutes, no baseline regressions)
- `git diff --check` and Python compilation passed
- verified public/private delivery, whole-date availability wording, confirmation ownership/expiry, duplicate handling, backend failures, routing separation, catalog linting, and evaluation cases

## Dependency and rollout

Depends on backend PR [MLAI-AUS-Inc/mlai-backend#710](https://github.com/MLAI-AUS-Inc/mlai-backend/pull/710).

Deploy backend migration/API first, deploy Roo second, smoke-test availability/booking/replay/cancellation/refund, then enable the feature flag.

### CODE REVIEW

**Comment by Light-code11** [comment]
Local verification after rebasing onto current `main`:

- meeting-room interaction, client, Slack action, router, catalog, and routing-evaluation tests: **36 passed**
- router catalogs remain below the 6,000-token gate (largest context approximately 5,944 tokens)
- branch is mergeable with `main`

GitHub's `security-checks` job did not start. Its check-run annotation is: “The job was not started because your account is locked due to a billing issue.” This is an organisation GitHub Actions/billing blocker rather than a test failure. The check should be rerun after billing access is restored.

**Comment by Light-code11** [comment]
Rebased onto current `main` and updated whole-date availability so Roo lists unavailable intervals and explicitly states that all other times are available. Local validation passed (511 CI-equivalent tests; 38 focused meeting-room/router tests; 249 routing cases with 0 misroutes), and the GitHub security check is green. Please merge backend PR MLAI-AUS-Inc/mlai-backend#710 and confirm its main deployment before merging this Roo PR.

File: roo-standalone/roo/main.py [hunk 1]
```diff
@@ -5091,6 +5192,22 @@ async def slack_actions(
...
+        MEETING_ROOM_CANCEL_ACTION_ID,
+    }:
+        asyncio.create_task(
+            _handle_meeting_room_action(
+                settings=settings,
+                action_id=str(action_id),
+                action_value=str(actions[0].get("value") or ""),
+                actor_user_id=str(user_id or ""),
+                channel_id=str(channel_id or ""),
+                message_ts=str(thread_ts or ""),
```
Comment (callumholt) [inline roo-standalone/roo/main.py]: **`message_ts` is the thread parent, not the button message.**

At line 5145 `thread_ts = message.get("thread_ts") or message.get("ts")`, and skill results are posted as threaded replies (`thread_ts = event.get("thread_ts") or event.get("ts")` in the DM/mention handlers). So for a confirmation card posted in reply to the member's DM, `message["thread_ts"]` is the **member's** message ts, not the card's.

Concretely: member DMs "book the meeting room tomorrow at 2pm" → Roo posts the confirm card as a reply in that thread → member clicks Confirm → `_handle_meeting_room_action` calls `chat_update(ts=<member's message ts>)`, which Slack rejects (`cannot_update_message`), the `except` fires and the outcome is posted as a brand-new message. The card keeps its live Confirm/Cancel button afterwards. If the thread parent happens to be a bot message instead, `chat_update` succeeds and overwrites an unrelated Roo message while the card still shows a clickable button.

Use the container/message ts instead, e.g. `payload.get("container", {}).get("message_ts") or message.get("ts")`.

File: roo-standalone/roo/skills/executor.py [hunk 2]
```diff
@@ -433,6 +450,262 @@ async def _execute_committee_candidate_emails(
...
+                message = self._format_meeting_room_availability(availability)
+                return self._deliver_meeting_room_response(
+                    user_id=user_id,
+                    channel_id=channel_id,
+                    message=message,
+                    action=action,
+                )
+
+            if action == "book_meeting_room":
+                starts_at, ends_at = resolve_meeting_room_interval(text, params)
```
Comment (callumholt) [inline roo-standalone/roo/skills/executor.py:699]: **No guard against booking a time already in the past.**

`resolve_local_date` happily returns today for `"today"`/a bare weekday name matching today, and `resolve_interval` never compares the resolved `starts_at` against `now`. At 4pm a member saying "book the meeting room today at 2pm" (or "book the room monday at 10am" on a Monday afternoon) gets a resolved interval in the past. The availability probe will most likely report it free (nothing is booked in the past), so Roo renders a Confirm card with a points cost; the failure only surfaces when the button is clicked, and `backend_error_message` has no mapping for a past-booking code, so the member sees the generic "I could not complete that meeting-room request."

Worth rejecting `starts_at <= now` client-side with a clear message before the availability call.

File: roo-standalone/roo/meeting_room_booking.py [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+
+
+def format_booking_result(result: dict) -> str:
+    booking = result.get("booking") or {}
+    starts_at = parse_backend_timestamp(booking.get("starts_at"))
+    ends_at = parse_backend_timestamp(booking.get("ends_at"))
+    room_name = str((booking.get("room") or {}).get("name") or "Meeting Room")
+    points = int(result.get("points_cost", booking.get("points_cost", 0)) or 0)
+    balance = result.get("remaining_balance")
+    replay = bool(result.get("already_booked"))
```
Comment (callumholt) [inline roo-standalone/roo/meeting_room_booking.py]: **Replay detection reads `already_booked`, but every other read of this response uses `created`.**

`_handle_meeting_room_action` logs `created={bool(result.get('created'))}` and the test fake returns `{"created": True, ...}` — so the backend's replay signal is `created: false`. Nothing else in the PR ever produces or consumes `already_booked` for this endpoint.

If the backend only returns `created`, a second click on the same Confirm button (same `client_request_id`, idempotent replay) reports "Your booking is confirmed." and re-states the charge, making it look like a second booking was made. Safer:

```suggestion
    replay = not bool(result.get("created", True))
```

File: roo-standalone/roo/meeting_room_booking.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+        "confirmation_expires_at": (issued_at + CONFIRMATION_TTL).isoformat(),
+    }
+    return json.dumps(payload, separators=(",", ":"), sort_keys=True)
+
+
+def build_cancel_action_value(*, owner_slack_user_id: str, booking_id: str) -> str:
+    return json.dumps(
+        {
+            "owner_slack_user_id": owner_slack_user_id,
+            "booking_id": str(UUID(str(booking_id))),
```
Comment (callumholt) [inline roo-standalone/roo/meeting_room_booking.py]: **Bare `ValueError` escapes the executor's exception handling.**

`UUID(str(booking_id))` raises a plain `ValueError`, not a `MeetingRoomInputError`. `_execute_meeting_room_booking` catches `MeetingRoomInputError`, `httpx.HTTPStatusError` and `MLAIBackendUnavailableError` — a plain `ValueError` is not among them (note `except MeetingRoomInputError` does not catch its own `ValueError` base).

So if `/meeting-rooms/my-bookings/` ever returns a booking whose `id` is not a UUID (an integer PK, a slug, or `null` on a partially-populated row), `cancellation_selection` blows up and the exception propagates out of the skill executor instead of producing a member-facing message. Wrap it and raise `MeetingRoomInputError`, or skip malformed rows.

File: roo-standalone/roo/main.py [hunk 5]
```diff
@@ -4788,6 +4798,97 @@ async def _review_admin_action(
...
+        from .slack_client import get_slack_client
+
+        get_slack_client().chat_update(
+            channel=channel_id,
+            ts=message_ts,
+            text=outcome,
+            blocks=blocks,
+        )
+    except Exception:
+        post_message(channel=channel_id, text=outcome, blocks=blocks)
```
Comment (callumholt) [inline roo-standalone/roo/main.py]: **Fallback `post_message` is outside any `try`, inside a fire-and-forget task.**

This whole function runs via `asyncio.create_task(...)` at line 5199 with no result awaited. `post_message` re-raises on Slack errors, so if `chat_update` fails (see the `message_ts` issue above) *and* the fallback post also fails — e.g. Slack rate limit, or `channel_id` empty — the exception surfaces only as an "exception was never retrieved" warning on GC. The member clicked Confirm, the booking was made and points charged, and they get no message at all.

Wrap the fallback in its own `try/except` that at least logs.

File: roo-standalone/skills/meeting_room_booking/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    description: Check availability.
+    params:
+      date: {type: string}
+      start_time: {type: string}
+  - name: book_meeting_room
+    description: Prepare booking confirmation.
+    params:
+      date: {type: string}
+      start_time: {type: string}
+      duration_hours: {type: integer}
```
Comment (callumholt) [inline roo-standalone/skills/meeting_room_booking/SKILL.md]: **No `end_time` param, so an explicit end time only survives if the member literally writes "from … to …".**

`resolve_interval` reads `params.get("end_time")`, but the schema never exposes it, so the router can never fill it. The fallback is `_natural_time_tokens`, whose range regex requires the literal keyword `from`:

```python
rf"\bfrom\s+(?P<start>{token})\s+(?:to|until|-)\s*(?P<end>{token})\b"
```

So "book the meeting room tomorrow 2pm-4pm" or "book the room tomorrow, 2pm until 4pm" produce `natural_end is None`. If the model also omits `duration_hours`, `raw_duration` falls through to `1` and the member is silently shown a **one-hour** confirm card for a two-hour request. Adding `end_time: {type: string}` to the `book_meeting_room` (and `check_room_availability`) params would close the gap.

File: roo-standalone/roo/main.py [hunk 7]
```diff
@@ -4788,6 +4798,97 @@ async def _review_admin_action(
...
+                    )
+                    outcome = format_meeting_room_cancellation_result(result)
+                    print(
+                        "MEETING_ROOM_BOOKING_CANCELLED "
+                        f"booking_id={value['booking_id']} "
+                        f"cancelled={bool(result.get('cancelled'))}"
+                    )
+        except MeetingRoomInputError as exc:
+            outcome = exc.message
+        except Exception as exc:
```
Comment (callumholt) [inline roo-standalone/roo/main.py:5167]: **"Nothing was changed" can be false after a `/book/` call.**

`book_meeting_room` is a POST with `transport_retries=1` and a 15s timeout. A read timeout or connection drop *after* the backend committed the booking and deducted points raises `MLAIBackendUnavailableError`, which lands here and produces `backend_error_message`'s default: "I could not complete that meeting-room request. Nothing was changed." The member is told nothing happened while they have been charged and the room is reserved.

Since the request is idempotent on `client_request_id`, a safer wording for the book path on transport/5xx failures is something like "I could not confirm the result — check `show my meeting room bookings` before trying again", rather than asserting nothing changed.

File: roo-standalone/roo/tests/test_router_v2.py [hunk 8]
```diff
@@ -103,10 +103,10 @@ def test_skill_tool_embeds_routing_block_and_action_enum():
         if tool["function"]["name"] == "mlai-points"
     )
     description = tool["function"]["description"]
-    assert "Use when:" in description
-    assert "Do NOT use when:" not in description or "avoid" in description.lower()
-    assert "Examples:" in description
-    assert "Counter-examples" in description
+    assert "Use:" in description
+    assert "Avoid:" not in description or "avoid" in description.lower()
```
Comment (callumholt) [inline roo-standalone/roo/tests/test_router_v2.py]: This assertion is vacuous and can never fail: if `"Avoid:"` is in the description then `"avoid"` is necessarily in `description.lower()`, and if it isn't, the left disjunct is true. It asserts nothing about the new `Avoid:` prefix. Either drop it or make it a real check:

```suggestion
    assert "Avoid:" in description
```

**Comment by callumholt** [comment]
Reviewed alongside #209 and #210. Eight inline comments above. A few things that are about the shape of the PR rather than any one line:

**The `router.py` change should be its own PR.** It rewrites the routing-prompt scaffolding for *every skill in the catalog* — `Use when:` → `Use:`, `Do NOT use when:` → `Avoid:`, `Counter-examples (pick the other tool):` → `Not:`, `Examples:` → `Eg:` — purely to reclaim tokens so the new skill fits the budget. That is a global change to how the model is instructed to route, bundled into a feature PR that is otherwise fully gated behind `MEETING_ROOM_BOOKING_ENABLED`. The router change is not gated by anything.

The 249-case deterministic eval passing does not really cover it: those cases are blessed and deterministic, and this change targets exactly the LLM-generalisation layer they are designed not to depend on. `Do NOT use when` → `Avoid` is a genuine reduction in instruction force, and `Counter-examples (pick the other tool)` → `Not:` drops the explicit instruction about what to do with the counter-example. Splitting it out means it can be evaluated in full mode on its own merits, and reverted independently if routing quality moves.

**Two input-handling gaps not covered inline:**

- `resolve_interval` rolls a reversed interval forward a day: `if ends_at <= starts_at: ends_at = _local_datetime(local_date + timedelta(days=1), end_time)`. A typo like "2pm to 1pm" becomes a 23-hour booking, rendered as such in the priced confirmation card. The backend's four-hour limit would catch it, but rejecting it at parse time gives a much better message.
- `_parse_time_value` raises `ambiguous_time` for any bare hour ≤ 12, so `09:00` — unambiguous 24-hour input — gets "Is the start time AM or PM?". Treating a `HH:MM` form (or a leading zero) as 24-hour would fix it.

**Minor:** `list_meeting_rooms` has no production caller — only the test suite calls it.

**Credit where due:** the DST handling in `_local_datetime` is genuinely careful (both the non-existent hour and the twice-occurring hour are rejected rather than silently resolved), `_add_actual_hours` correctly measures elapsed rather than wall-clock hours across a transition, the config validation fails closed on surface and credentials, and adding the test file to the PR security gate in `deploy.yml` is the right instinct.

Note mlai-backend#710 is still open, so this cannot deploy yet regardless.

File: roo-standalone/roo/main.py (hunk 1 above)
Comment (Light-code11) [reply roo-standalone/roo/main.py]: Fixed. Meeting Room actions now use Slack container.message_ts, falling back only to message.ts, rather than using the thread parent. The signed-action regression uses distinct parent, message, and container timestamps and asserts that the confirmation card timestamp is scheduled for update.

File: roo-standalone/roo/skills/executor.py (hunk 2 above)
Comment (Light-code11) [reply roo-standalone/roo/skills/executor.py:699]: Fixed. Resolved intervals now reject starts at or before the current time before Roo calls availability. A regression covers a 2pm request made at 4pm and verifies the clear future-time error.

File: roo-standalone/roo/meeting_room_booking.py (hunk 3 above)
Comment (Light-code11) [reply roo-standalone/roo/meeting_room_booking.py]: Fixed. The formatter now treats created=false as an idempotent replay when created is present, with already_booked retained only as a compatibility fallback. A regression covers a replay response containing created=false and no already_booked field.

File: roo-standalone/roo/meeting_room_booking.py (hunk 4 above)
Comment (Light-code11) [reply roo-standalone/roo/meeting_room_booking.py]: Fixed. Booking UUID normalization now converts malformed values into MeetingRoomInputError. Cancellation selection skips malformed rows when valid rows remain and returns a member-facing invalid-response message when none are usable; both cases have regression coverage.

File: roo-standalone/roo/main.py (hunk 5 above)
Comment (Light-code11) [reply roo-standalone/roo/main.py]: Fixed. The fallback post is wrapped in its own exception handler and emits a bounded delivery-failure event containing only the action and exception class names. A regression proves simultaneous chat_update and fallback failures do not escape the background task.

File: roo-standalone/skills/meeting_room_booking/SKILL.md (hunk 6 above)
Comment (Light-code11) [reply roo-standalone/skills/meeting_room_booking/SKILL.md]: Fixed. end_time is now declared for both availability and booking actions. The natural parser also preserves 2pm-4pm and 2pm until 4pm without requiring the word from. Schema and parameterized parsing regressions cover all three paths.

File: roo-standalone/roo/main.py (hunk 7 above)
Comment (Light-code11) [reply roo-standalone/roo/main.py:5167]: Fixed. Once a booking or cancellation mutation begins, transport and 5xx failures now say Roo could not confirm the result and instruct the member to check show my meeting room bookings before retrying. Known 4xx errors retain their specific message. A backend-unavailable regression verifies that Nothing was changed is not emitted.

File: roo-standalone/roo/tests/test_router_v2.py (hunk 8 above)
Comment (Light-code11) [reply roo-standalone/roo/tests/test_router_v2.py]: Addressed in separate prerequisite PR #232. The test fixture now includes an avoid_when rule and directly asserts the rendered Use, Do NOT, Examples, and negative-example guidance, so the assertion is no longer vacuous. PR #225 is stacked on #232 and contains no router.py or router-test diff.

**Comment by Light-code11** [comment]
@callumholt I addressed the complete review and pushed a rebased update.

The global router scaffolding is now isolated in #232, and #225 is stacked on that branch with no router.py or router-test files in its comparison. The prerequisite keeps explicit Do NOT guidance and explicit alternative-tool direction rather than the weaker Avoid/Not wording.

The remaining summary items are covered as follows:
- 2pm to 1pm is rejected locally as an over-four-hour probable reversal, while valid 11pm to 1am bookings still cross midnight.
- 09:00 is treated as unambiguous 24-hour input.
- list_meeting_rooms was removed because it had no production caller.
- every inline thread has a focused regression test.

Validation completed:
- #232 CI-equivalent suite: 493 passed
- #225 CI-equivalent suite: 527 passed
- meeting-room/router regressions: 56 passed
- deterministic routing: 249 cases, 0 misroutes, no baseline regressions
- full repository: 876 passed; 20 unrelated failures reproduce on current main

A live LLM routing run was not possible locally because no real OpenAI API key is available.

**Comment by Light-code11** [comment]
Updated the Roo skill for the requested booking rules and admin flow.

- accepts `an hour and a half`, `1.5 hours`, and `90 minutes`
- enforces a 2-hour maximum and hour/half-hour boundaries
- previews 90 minutes as 1.5 hours costing 2 integer Roo Points
- full Points Admins can book one explicitly tagged member; the target is charged and the admin is not
- labeled Slack mentions retain the verified target ID through normalization
- confirmation binds the actor, target, request ID, expiry, and previewed price
- invalid/fractional action payloads are rejected before a backend call

Validation: 539 CI-equivalent tests passed, 52 focused meeting-room/catalog/routing tests passed, the 249-case deterministic routing evaluation had 0 misroutes and no baseline regression, and the catalog remains below its 6,000-token limit. The global router scaffolding remains isolated in #232.

**Comment by Light-code11** [comment]
Rebuilt #225 linearly on the corrected #232 head and completed the remaining review hardening.

- cancellation now has a model-visible routing example as well as deterministic coverage
- the over-40 cancellation guard has a focused regression and asks for a date before constructing Slack buttons
- meeting-room button-handler tasks are retained until completion instead of relying on an unreferenced fire-and-forget task
- duplicate Slack delivery still schedules only one action

Validation: 63 focused tests passed, the 541-test CI-equivalent suite passed, and 249 deterministic cases produced 0 misroutes with no baseline regression. The full suite is 890 passed / 20 failed; the same 20 failures reproduce on prerequisite #232. The Victor catalog is approximately 5,997.25 tokens, below the 6,000-token gate. A live OpenAI-backed eval could not be run locally because no real API key is available.

**Comment by Light-code11** [comment]
Addressed the latest validation feedback on #225.

- newly created admin bookings now privately notify the member whose Roo Points were charged
- the target DM states who booked them, the room/time, the charge, and how to view or cancel the booking
- the admin confirmation reports whether target delivery succeeded; delivery failure does not roll back or misreport the completed booking
- idempotent `created=false` replays do not send duplicate target DMs
- `list_my_room_bookings` is restored as a model-visible routing example while the Victor catalog remains below budget
- labeled-mention compatibility now covers Meeting Room booking, connect-users, and Content Factory delegation

Validation completed:

- 74 focused Meeting Room/agent/router/catalog tests passed
- 552 CI-equivalent tests passed
- 249 deterministic routing cases, 0 misroutes, no baseline regressions
- full repository: 894 passed / the same 20 pre-existing failures reproduced on #232
- catalog size: approximately 5,343.5 public and 5,993.5 Victor tokens
- Python compilation and `git diff --check` passed

The only remaining item from the feedback is the live OpenAI-backed routing evaluation. No real `OPENAI_API_KEY` is available in this environment, so I did not overwrite the deterministic baseline or claim that run passed.

**Comment by Light-code11** [comment]
Addressed the adversarial review in `1d70ef3`.

The two merge-blocking failure modes now have durable recovery:

- valid booking/cancellation clicks are persisted before Slack acknowledgement, leased by one worker, recovered after restarts, and drained on graceful shutdown
- an admin booking whose backend response is lost is replayed with the same request ID and still notifies the charged member
- target notification completion is stored independently, so a failed admin-card update cannot cause repeated target DMs
- target DMs use a stable Slack `client_msg_id`

The remaining findings are also fixed:

- tagged-member errors are target-aware
- bare `14:00` availability requests retain their exact interval
- labelled Roo mentions are removed correctly
- CI explicitly includes agent routing, the new outbox suite, and the deterministic routing gate; the workflow is configured for pull requests to any base branch

Validation:

- 82 focused tests passed
- 572 workflow-equivalent tests passed
- 249 deterministic routing cases, 0 misroutes
- full suite: 911 passed / the same 20 prerequisite-branch baseline failures
- Python compilation, workflow YAML validation, and `git diff --check` passed

Docker was unavailable locally, so Compose validation remains for GitHub CI. This PR is still stacked on #232; merge #232 and retarget #225 to `main` before merge. Backend #710 must deploy before Roo.

**Comment by callumholt** [comment]
## PR review

I reviewed the Meeting Room parsing, privacy boundaries, Slack action verification, durable outbox/retry lifecycle, backend client contract, admin-target flow, configuration gates, and the companion backend API at commit `1d70ef30d65ca953bebfba4696e8c1e77d3728f3`.

### Blocking finding

**Durable retries stop working after the confirmation TTL expires.**

`_handle_meeting_room_action` checks `meeting_room_confirmation_expired(value)` before calling `book_meeting_room`, including when it is processing a durable retry with `retry_failures=True`. Consider this supported failure path:

1. The backend commits the booking and deducts points, but Roo loses the response.
2. Roo persists/releases the outbox action for an idempotent replay.
3. Roo is unavailable or unable to process the action until after the ten-minute confirmation TTL.
4. On restart, Roo returns the local “confirmation expired” outcome and completes the outbox row without replaying `client_request_id`.

The booking remains real and charged, while Roo can fail to recover the result or notify an admin-booked target. This defeats the stated restart/uncertain-commit recovery guarantee. The companion backend already supports the correct behavior: it looks up an existing `client_request_id` before validating confirmation expiry, so an idempotent replay can safely recover after expiry.

For durable processing, let the backend decide expiry (or persist whether a mutation was attempted and bypass the local expiry check only for uncertain retries). Keep the local check for first/non-durable attempts if desired. Please add a regression covering an expired queued action whose backend replay returns `created=false`, including target notification recovery.

### Other review results

- Privacy-sensitive availability, booking, balance, cancellation, and target details remain on the DM path.
- Actor/target binding is rechecked from the signed Slack action and by the strict-key backend API.
- Stable request IDs, price binding, target-aware errors, lease ownership, cancellation idempotency, and separate target-notification persistence are otherwise well designed.
- Focused suites pass when isolated by process: **68 Meeting Room/outbox tests** and **14 agent/router/catalog tests** (**82 total**).
- Deterministic routing evaluation: **249 cases, 0 misroutes, no baseline regressions**.
- GitHub `security-checks`: passed (**572 tests**).
- Python compilation and `git diff --check`: passed.
- I inspected backend PR #710 and its migration/API contract read-only; I did not run or apply any database migration.

### Dependencies

This PR remains stacked on open prerequisite #232. Backend PR #710 is also open and must be merged, migrated, and deployed before Roo is deployed; the feature flag should remain disabled until the documented smoke tests pass.

**Conclusion: Not safe or ready to merge yet. Fix the expired durable-replay gap, then merge/retarget after #232 and only roll out after backend #710 is deployed and verified.**

**Comment by Light-code11** [comment]
Addressed the latest blocker and completed another adversarial pass in cc53c3d.

The failure sequence from the latest review is now covered directly:
- persisted booking actions bypass the local confirmation-expiry check
- the backend receives the original client_request_id after the 10-minute TTL
- an idempotent created=false response recovers the final result
- an admin-booked target is still notified privately with the stable Slack client_msg_id

Additional state-machine and routing findings fixed:
- retry batches now lease each action immediately before processing, preventing later records from losing their lease behind a slow first request
- disabling the feature no longer drains uncertain queued mutations as completed; the queue remains untouched until re-enabled, while old buttons receive a non-mutating disabled response
- "show the meeting room calendar tomorrow" now routes to the private day-availability view rather than being excluded as external calendar work

Fresh validation:
- 575 CI-equivalent tests passed
- 85 focused Meeting Room/action/router/catalog tests passed
- 250 deterministic routing cases, 0 misroutes, no baseline regressions
- full repository: 914 passed / the same 20 unrelated prerequisite-branch failures
- Python compilation, git diff --check, undefined-name/syntax lint, and security static analysis passed with no medium/high findings
- backend #710 contract and green checks were re-verified read-only

A live OpenAI-backed routing eval remains unavailable because this environment has no real API key. The feature flag remains disabled by default, and backend #710 must deploy before Roo is enabled.

@callumholt please re-review the updated head.

**Comment by callumholt** [comment]
## PR re-review

I reviewed the changes since the prior review, from `1d70ef30d65ca953bebfba4696e8c1e77d3728f3` through `cc53c3d82a7c3729c3e9b61c4351e7f8c92ec206`, with emphasis on the expired durable-replay blocker.

### Prior blocker: resolved

Durable processing now bypasses only Roo's local confirmation-expiry check (`retry_failures=True`) and replays the stable `client_request_id` to the backend. This matches backend #710's ordering, where an existing idempotent result is recovered before confirmation expiry is evaluated. The new regression covers an expired queued admin booking whose original backend commit succeeded but response was lost, including recovery of the target DM with the stable Slack message ID.

The related changes are sound:

- Fresh/non-durable clicks still enforce the local ten-minute confirmation window.
- Disabling the feature stops new actions and pauses the retry worker without consuming previously queued uncertain mutations.
- Due actions are leased one at a time immediately before processing, avoiding lease expiry for later items while an earlier item is slow.
- The Meeting Room calendar wording now routes through the availability action without reintroducing generic calendar ownership.

### Verification

- Meeting Room action/booking suites: **71 passed, 0 failed**.
- Agent/router/catalog gates: **11 passed, 0 failed**.
- Deterministic routing evaluation: **250 cases, 0 misroutes, no baseline regressions**.
- GitHub `security-checks`: passed.
- Python compilation and `git diff --check`: passed.
- Backend PR #710 remains open with green checks. I inspected its contract/migration state read-only and did not run or apply any database migration.

This PR is still stacked on open #232 (`codex/router-prompt-budget`) rather than `main`, so merge it into its current base or land #232 and retarget/revalidate before merging to `main`. Backend #710 must be migrated and deployed before Roo is enabled.

**Conclusion: The code changes are safe and the prior blocker is fixed. This is ready to merge into its current stacked #232 base, but it is not ready for production rollout until #232 is landed/reconciled and backend #710 is merged, its migration is explicitly approved/applied, and the documented smoke tests pass.**

**Comment by Light-code11** [comment]
Addressed the latest adversarial review in `293e632`:

- bare `H:MM` times such as `2:30` now ask for AM/PM, while `09:30` and `14:00` remain unambiguous
- `half an hour` is parsed and rejected by the one-hour booking minimum instead of defaulting to one hour
- retries are bounded by attempt count and age, then parked in a terminal needs-attention state
- removed the ineffective `client_msg_id`/`duplicate_message` handling and its PR-body claim
- persisted target notification states prevent automatic duplicate DMs after an expired lease or unknown Slack outcome
- completed/failed button re-clicks replay the stored result instead of silently returning
- availability can inspect 30-minute to 24-hour windows through backend companion #710 while bookings remain capped at two hours
- an implicit same-day weekday whose time has passed rolls to the following week; explicit `today` and ISO dates do not
- added concurrent legacy SQLite schema-upgrade coverage

The persist-before-ack storage failure behavior remains deliberate: no mutation starts unless the click is durable, and the member can click again after Slack shows the failed action. This avoids acknowledging an action that Roo cannot recover.

Validation: 87 focused Meeting Room tests and 591 workflow-equivalent tests passed; compilation and `git diff --check` also passed.

@callumholt this is ready for another review when convenient.

**Comment by Light-code11** [comment]
Updated #225 for Small and Big Meeting Rooms in `92f5b00`.

- explicit Small/Big/Large language resolves from the verified user message
- unspecified availability checks both rooms privately
- unspecified bookings show private room-choice buttons
- durable first-choice-wins prevents duplicate or competing clicks from creating multiple confirmation flows
- the selected room, actor, interval, target, expiry, and stable booking request ID stay bound through confirmation
- room identity is revalidated against the backend before confirmation
- unknown Slack destinations no longer receive private room data inline

Fresh validation: 111 focused tests, 610 workflow-equivalent tests, 50/50 repeated simultaneous room-choice races, and 253 deterministic routing cases with 0 misroutes. @callumholt please re-review this updated head alongside backend #710.

**Comment by Light-code11** [comment]
Addressed the latest adversarial re-review in `d811519`.

- losing Small/Big clicks now receive a separate private message naming the room already selected
- that feedback uses `chat.postMessage` in the verified DM and never replaces the winning confirmation card
- exact Slack transport retries suppress repeated feedback; an endpoint regression verifies one real losing click plus one retry yields one message
- removed the unreachable choose-replay branch
- Roo credential/scope failures (`invalid_auth`, `token_revoked`, `not_authed`, `missing_scope`) now retry within the existing 20-attempt/24-hour bound
- target-side account/user-not-found failures remain terminal
- documented the per-room booking-ID/shared-selection-ID invariant
- word-number `for two hours` now parses correctly

Fresh validation: 119 focused tests, 617 workflow-equivalent tests, 50/50 repeated simultaneous-choice races, Python compilation, and diff checks passed.

@callumholt please re-review the updated head when convenient.

**Comment by callumholt** [comment]
## PR re-review

I reviewed the substantial changes since my last review, from `cc53c3d8` through `d8115192`, including the hardened time parsing and action delivery, Small/Big room selection, competing-click behavior, target-notification state machine, backend contract changes, and new fault-injection coverage.

The incremental implementation is internally coherent:

- a shared durable selection ID makes the first Small/Big choice win, while a losing real click receives private feedback without replacing the winning confirmation card;
- room, actor, interval, target, expiry, and booking request ID remain bound through the confirmation flow;
- target notifications now explicitly distinguish pending, sent, failed, and unknown-delivery states, avoiding automatic duplicate DMs after an abandoned sending lease;
- Roo credential/scope errors retry within the existing bounds, while target-specific permanent failures stop;
- the revised time parser and availability window handling cover the reported edge cases.

Focused Meeting Room action/booking verification passed: **111 passed, 0 failed**. Two legacy SQLite schema-upgrade tests were deliberately excluded; I did not run or apply any database migration. Python compilation, `git diff --check`, and the GitHub security check passed.

However, repository state supersedes that isolated result: #225 was merged and then fully rolled back by emergency PR #235 while reported issues are investigated. The Meeting Room code is no longer on `main`, and companion backend #710 remains open. A green focused suite is not sufficient evidence to re-land code that was explicitly rolled back for unresolved operational concerns.

**Conclusion: #225 has already been merged and reverted, and it is not safe or ready to re-merge/re-land until the reason for emergency rollback #235 is documented, reproduced or otherwise resolved, and the complete Roo/backend rollout is revalidated.**
